056-kafka-roller-kraft.md

```diff
@@ -0,0 +1,158 @@
+- Mark a node as caught up if `leaderLastCaughtUpTimestamp - replicaLastCaughtUpTimestamp < controllerQuorumFetchTimeoutMs`. This will exclude the controller KafkaRoller is currently considering to roll, because we are trying to check whether the quorum would stay healthy when this controller is rolled.
+- Count each controller node that is caught up, including the leader (`numOfCaughtUpControllers`).
+- Can roll if: `numOfCaughtUpControllers >= ceil((double) totalNumOfControllers / 2)`.
+
+> NOTE: Until [KIP-919](https://cwiki.apache.org/confluence/display/KAFKA/KIP-919%3A+Allow+AdminClient+to+Talk+Directly+with+the+KRaft+Controller+Quorum) is implemented, KafkaRoller cannot create an admin connection to the controller directly to describe its configuration or the quorum state. Therefore KafkaRoller checks the desired configurations to get the value of `controller.quorum.fetch.timeout.ms` and creates admin client connection to the brokers for the quorum check. If KafkaRoller cannot connect to any of the brokers after the maximum attempts to retry, the controllers will be marked as failed to reconcile because of not being able to determine the quorum health. KafkaRoller would then try to reconcile the brokers, which may help restoring admin client connections to them. In this scenario, the reconcilation will be completed with failure and reported to the operator. The controllers quorum check will be retried in the next round of reconcilation.
+
+
+#### Separate brokers and controllers
+For controller-only:
+ 1. Does not force roll a controller performing log recovery
+ 2. Force rolls a broker if the pod is in stuck state
+ 3. Attempts to connect an admin client to any of the brokers, if cannot connect, skip checks in `4` and `5` 
+ 4. Rolls a controller if the controller configuration has changed
+ 5. Does not roll a controller if doing so would take the number of caught-up controllers (inc leader) to less than majority of the quorum
+
+For broker-only:
+ 1. Does not force roll a broker performing log recovery
+ 2. Force rolls a broker if the pod is in stuck state
+ 3. **NEW** Attempts to connect an admin client to a broker and if it can't connect rolls the broker 
+ (the order of the rollability checks will change, we will attempt admin client connection and force roll only if the pod is not stuck)
+ 4. Rolls a broker if the broker configuration has changed and cannot be updated dynamically
+ 5. Does not roll a broker if doing so would take the in-sync replicas count below `min.insync.replicas`
+
+
+#### Combined mode
+1. Does not force roll a combined node performing log recovery
+2. Force rolls a combined node if the pod is in stuck state
+3. Attempts to connect an admin client to any of the brokers, if cannot connect, skip checks in `4`, `5` and `6`
```

Comment:
skip checks but then? rolls?

Reply:
This was actually incorrect , sorry! I have corrected this part.